### PR TITLE
[None][feat] Kimi K3: unlock the CUTEDSL MoE backend for NVFP4 SiTU

### PR DESCRIPTION
**Pull request by xguannv**
## Description

Enable explicitly selected CUTEDSL routed experts for Kimi K3 NVFP4 on SM100/SM103. The Blackwell kernel already implements SiTU in the rebased main; this change enables it through the MoE capability declaration and K3 backend selection, forwards both soft-caps, and rejects an explicit CUTEDSL request that cannot be served instead of silently selecting CUTLASS.

The Blackwell custom op and fake schema use `Optional[float] = None` directly. SiTU constants must be finite and positive. SM107 end-to-end SiTU integration remains gated; existing Rubin ops retain their separate legacy sentinel API.

## Review coverage

- Compare the NVFP4 output with a SiTU reference and a **plain SwiGLU** negative reference, using dequantized checkpoint weights and a SiTU relative-L2 bound of **0.20**.
- Derive backend admission candidates from `BACKEND_FAMILY` and check exact agreement with the K3 allow-list.
- Exercise the real `KimiK3MoERuntime` / resolver call chain with CUTEDSL ineligible and CUTLASS eligible; require an explicit rejection, with an allowed-degradation positive control.
- At fixed outer/FC1 tactics, independently change each soft-cap across `(4,25)`, `(2,25)`, `(4,10)`, `(2,10)` and verify distinct runner identities and numerical outputs.
- Register the regressions in `l0_b300.yml`. Keep the numerical discriminator's CI selection focused on CUTEDSL as requested; retain the existing TRTLLM selector. Update the upstream Blackwell direct-op test to use `None` for absent SiTU parameters.

## Validation

Current PR head: `e04607744e7571a9817ce4bc0add0434591366f5`. GPU validation ran at `74a480e98701ba8423151f68503edd5e6db31e7b`; the only subsequent change restores three example files to the base version. Implementation, unit tests, and CI selectors are unchanged. The DEP16 quickstart result below used the earlier example launcher with explicit backend selection. The branch was rebased onto `63e64e5bdb447af0eb12b5b6ca6a185483131783`.

OCIAGA GB300/SM103, image `trtllm-ociaga-gb300-k3-base-nsys2025.5-20260916.sqsh`; `libtensorrt_llm.so` SHA256 `3eac19a923ae4df65df65bd63d03a64d5716f5ce298424433738eeabc78b5eeb`. Build job **802376** used `ad95dd32a303`; the GPU-validated candidate differs only in the CI selector YAML.

| Check | Result |
| --- | --- |
| Local pre-commit / CI-selector validation / DCO | Passed |
| K3 SiTU tests, job 802479 | 83 passed, 1 existing strict xfail |
| Blackwell direct-op tests, separate process, job 802479 | 6 passed |
| MoE backend suite, separate process, job 802479 | 430 passed, 166 skipped |
| DEP16 quickstart, job 805968 | 4/4 expected responses |
| GSM8K, job 806041 | **96.3609 ± 0.5158% strict; 96.5125 ± 0.5053% flexible** |

GSM8K used `nvidia_kimi-k3-nvfp4/hf/hf-f8c5234_orig`, 4 nodes × 4 GPUs in one NVLink segment, DEP16, batch size 8, KV fraction 0.35, 5-shot, all **1319/1319** questions, and a 256-token output limit. Nodes: `nvl72d136-T[02,07-09]`.

The **same GSM8K run** recorded both `CuteDslFusedMoE::run_moe_nvfp4` and `trtllm::cute_dsl_nvfp4_gather_grouped_gemm_act_fusion_blackwell` on every rank 0–15. These are runtime call-path records, emitted once per rank during warmup, rather than per-question kernel counts. There were no `trtllm::fused_moe::gemm1` records or resolver declines. The four soft-cap unit cases had relative L2 errors 0.164376 / 0.167295 / 0.165008 / 0.167400.

These are local cluster validation results, not a claim that official PR GPU CI has passed. No performance improvement or backend ranking is claimed.

## PR Checklist

- [x] Description and tests cover the final change.
- [x] Reviewed the coding guidelines and applicable PR requirements.
- [x] CI selectors updated.
- [x] No new dependencies or ownership changes.

### CODE REVIEW

**Review by leslie-fang25 — APPROVED** [review]

File: tests/unittest/_torch/moe/test_kimi_k3_situ_moe.py [hunk 1]
```diff
@@ -1493,7 +1660,7 @@ def _swiglu_reference_moe(x, router_logits, routing_method, w1, w2, w3, alpha, b
 
 
 @nvfp4_moe_supported
-@pytest.mark.parametrize("moe_backend", ["CUTLASS", "TRTLLM"])
+@pytest.mark.parametrize("moe_backend", _NVFP4_SITU_BACKENDS)
```
Comment (sunnyqgg) [inline tests/unittest/_torch/moe/test_kimi_k3_situ_moe.py:1663]: CUTEDSL uses plain SwiGLU, but this test’s negative reference is SwigluBias(alpha=4, beta=25). Please compare against plain SwiGLU and assert an error bound; the current cosine ordering accepts both plain-SwiGLU output and incorrectly scaled SiTU output.

File: tests/unittest/_torch/moe/test_moe_backend.py [hunk 2]
```diff
@@ -2537,6 +2539,30 @@ def test_nvfp4_fc1_row_alignment_gate(
         assert verdict.reject_reason is not MoERejectReason.SHAPE_UNALIGNED
 
 
+@pytest.mark.parametrize(
+    "backend_cls",
+    [CutlassFusedMoE, CuteDslFusedMoE],
+    ids=["cutlass", "cutedsl"],
+)
+def test_situ_survives_resolution_not_just_construction(backend_cls):
```
Comment (sunnyqgg) [inline tests/unittest/_torch/moe/test_moe_backend.py:2547]: Please wire this test into CI. Existing GPU selectors do not select it, and CPU collection excludes this file because it has no pytest.mark.cpu_only marker.

File: tensorrt_llm/_torch/custom_ops/cute_dsl_custom_ops.py [hunk 3]
```diff
@@ -78,6 +78,29 @@ def _canonicalize_swiglu_limit_scalar(swiglu_limit_scalar: float) -> float:
     return float("inf") if swiglu_limit_scalar < 0 else swiglu_limit_scalar
 
 
+#: Sentinel for "this layer is not SiTU". A torch custom op schema cannot carry
```
Comment (sunnyqgg) [inline tensorrt_llm/_torch/custom_ops/cute_dsl_custom_ops.py:81]: PyTorch custom-op schemas support Optional[float] = None. Could we use that directly and remove the sentinel, canonicalization helper, and sentinel-only test?

File: tests/unittest/_torch/moe/test_kimi_k3_situ_moe.py [hunk 4]
```diff
@@ -591,6 +651,80 @@ def test_kimi_k3_routed_config_logs_megamoe_capacity_override(monkeypatch):
...
+    it is how CUTEDSL stayed shut after its kernels grew the epilogue.
+    """
+    from tensorrt_llm._torch.moe.fused_moe.fused_moe_cute_dsl import CuteDslFusedMoE
+    from tensorrt_llm._torch.moe.fused_moe.fused_moe_cutlass import CutlassFusedMoE
+    from tensorrt_llm._torch.moe.fused_moe.fused_moe_trtllm_gen import TRTLLMGenFusedMoE
+    from tensorrt_llm._torch.moe.fused_moe.mega_moe.mega_moe_cute_dsl import MegaMoECuteDsl
+    from tensorrt_llm._torch.moe.fused_moe.mega_moe.mega_moe_deepgemm import MegaMoEDeepGemm
+    from tensorrt_llm._torch.utils import ActivationType
+
+    declares_situ = {
```
Comment (sunnyqgg) [inline tests/unittest/_torch/moe/test_kimi_k3_situ_moe.py:669]: This dictionary duplicates the backend list and checks only inclusion, not the exact agreement claimed by the docstring. Consider simplifying this to the CUTEDSL admission regression or deriving candidates from the existing backend registry.

**Review by sunnyqgg — COMMENTED** [review]
We rarely use the CUTLASS backend for kimi k3, so could we keep this test focused on CUTEDSL and remove the CUTLASS case to avoid spending CI resources on it? thanks

**Review by rosong11 — APPROVED** [review]

**Review by chzblych — APPROVED** [review]

**Review by Wanli-Jiang — APPROVED** [review]

**Review by mikeiovine — APPROVED** [review]
